### PR TITLE
feat: `@effect/platform` adapter

### PR DESCRIPTION
**Pull request by juliusmarminge**

### CODE REVIEW

File: packages/uploadthing/src/effect-http.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  );
+  const getBuildPerms = buildPermissionsInfoHandler<TRouter>(opts);
+
+  return Http.router.empty.pipe(
+    Http.router.get(
+      "/",
+      Http.response.json(getBuildPerms(), {
+        headers: Http.headers.fromInput({
+          "x-uploadthing-version": UPLOADTHING_VERSION,
+        }),
```
Comment (juliusmarminge) [inline packages/uploadthing/src/effect-http.ts:53]: can we provide this on a router level? All responses we return from this router should have this header. we're currently setting it 3 times

File: examples/backend-adapters/server/src/effect.ts [hunk 2]
```diff
@@ -1,46 +1,59 @@
-import { HttpServer } from "@effect/platform";
-import { BunHttpServer, BunRuntime } from "@effect/platform-bun";
-import { Effect, Layer } from "effect";
+import "dotenv/config";
```
Comment (datner) [inline examples/backend-adapters/server/src/effect.ts:1]: I don't think this is actually needed tbh

File: examples/backend-adapters/server/src/effect.ts [hunk 3]
```diff
@@ -1,46 +1,59 @@
...
+        status: 204,
+        headers: Http.headers.fromInput(corsHeaders),
+      });
+    }
+
+    const response = yield* app;
+
+    return response.pipe(Http.response.setHeaders(corsHeaders));
+  }),
+);
```
Comment (datner) [inline examples/backend-adapters/server/src/effect.ts:42]: Every time I forget how simple cors is 🤦🏻

File: examples/backend-adapters/server/src/effect.ts [hunk 4]
```diff
@@ -1,46 +1,59 @@
...
-  BunRuntime.runMain,
+const app = router.pipe(
+  cors,
+  Http.server.serve(Http.middleware.logger),
+  Http.server.withLogAddress,
 );
+
+const port = +(process.env.PORT || "3000");
+
+const ServerLive = NodeHttpServer.server.layer(() => createServer(), { port });
```
Comment (datner) [inline examples/backend-adapters/server/src/effect.ts:57]:
```ts
const Port = Config.integer("PORT").pipe(Config.withDefault(3000))
const ServerLive = Layer.unwrapEffect(
  Effect.map(Port, port => NodeHttpServer.server.layer(() => createServer(), { port })
)

File: packages/uploadthing/src/effect-http.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+) => createBuilder<MiddlewareArgs, TErrorShape>(opts);
+
+export const createRouteHandler = <TRouter extends FileRouter>(
+  opts: RouteHandlerOptions<TRouter>,
+): Http.router.Router<Http.body.BodyError, never> => {
+  incompatibleNodeGuard();
+
+  const requestHandler = buildRequestHandler<TRouter, MiddlewareArgs>(
+    opts,
+    "effect-http",
```
Comment (datner) [inline packages/uploadthing/src/effect-http.ts:43]: note that `effect-http` [is a real thing](https://github.com/sukovanej/effect-http), we refer to the more low-level http server utilities as `@effect/platform`

File: packages/uploadthing/src/effect-http.ts (hunk 1 above)
Comment (datner) [reply packages/uploadthing/src/effect-http.ts:53]: Yes
```ts
const addUploadThingHeader = Http.middleware.make(
  Effect.map(Http.response.setHeader("x-uploadthing-version", UPLOADTHING_VERSION)),
);

Http.router.empty.pipe(
  // ...
  Http.router.use(addUploadThingHeader)
)
```
Technically `middleware.make` is just an `identity` type helper.
a `Middleware` is `<E,R>(app: App.Default<E,R>) => App.Default<any, any>`, and `App.Default` is just an alias for `Effect<ServerResponse, E,R>`. Meaning the above is just shorthand.

The `.use` would affect everything within the given `Router`. In a pipe remember that the composition is technically inverse so everything above the `.use` would be affected but not anything below it.
```ts
Http.router.empty.pipe(
  Http.router.get("a", A),
  Http.router.get("b", B),
  Http.router.get("c", C),
  Http.router.use(addUploadThingHeader)
)

// is literally
Http.router.use(addUploadThingHeader)(
  Http.router.get("c", C)(
    Http.router.get("b", B)(
      Http.router.get("a", A)(
        Http.router.empty
      )
    )
  )
)
```

File: packages/uploadthing/src/effect-http.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    ),
+    Http.router.post(
+      "/",
+      Effect.flatMap(Http.request.ServerRequest, (req) =>
+        requestHandler({
+          /**
+           * TODO: Redo this to be more cross-platform
+           * This should handle WinterCG and Node.js runtimes,
+           * unsure about others...
+           * Perhaps we can use `Http.request.ServerRequest` internally?
```
Comment (datner) [inline packages/uploadthing/src/effect-http.ts:64]: This would make it cross-platform indeed 💯

File: packages/uploadthing/src/effect-http.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+          Effect.provideService(FetchContext, {
+            fetch: opts.config?.fetch ?? globalThis.fetch,
+            baseHeaders: {
+              "x-uploadthing-version": UPLOADTHING_VERSION,
+              // These are filled in later in `parseAndValidateRequest`
+              "x-uploadthing-api-key": undefined,
+              "x-uploadthing-be-adapter": undefined,
+              "x-uploadthing-fe-package": undefined,
+            },
+          }),
```
Comment (datner) [inline packages/uploadthing/src/effect-http.ts:94]: continuing from our `Client` implementation:
```ts
Effect.provideServiceEffect(
  HttpClient.Client,
  HttpClient.mapRequest(
    HttpClient.fetch,
    ClientRequest.setHeader("x-uploadthing-version", UPLOADTHING_VERSION),
  ),
),
Effect.provideService(HttpClient.Fetch, opts.config?.fetch ?? globalThis.fetch),
```


Note though, currently this means that you must use `fetch`, `Undici` and `Agent`s are not supported. But you might want to support them

File: docs/src/pages/api-reference/server.mdx [hunk 8]
```diff
@@ -6,7 +6,7 @@ The helper function to create an UploadThing instance. MAKE SURE YOU IMPORT IT
 FROM THE RIGHT PLACE. The export name ensures your file routes' `middleware`
 functions are typed correctly.
 
-<Tabs items={["Next App Router", "Next Pages Dir", "SolidJS", "Express", "Fastify", "H3"]}>
+<Tabs items={["Next App Router", "Next Pages Dir", "SolidJS", "Effect HTTP","Express", "Fastify", "H3"]}>
```
Comment (datner) [inline docs/src/pages/api-reference/server.mdx:9]: `Effect Platform`? idk, ask the other folk over at effect how they wanna officiate this

File: examples/backend-adapters/server/src/effect.ts (hunk 2 above)
Comment (juliusmarminge) [reply examples/backend-adapters/server/src/effect.ts:1]: Tax doesnt load process.env automatically I dont think

File: examples/backend-adapters/server/src/effect.ts (hunk 4 above)
Comment (juliusmarminge) [reply examples/backend-adapters/server/src/effect.ts:57]: Does effect/Config work with process.env, import.meta.env etc? Wondering if we could replace std-env with it?

File: packages/uploadthing/src/effect-http.ts (hunk 7 above)
Comment (juliusmarminge) [reply packages/uploadthing/src/effect-http.ts:94]: cool - the client stuff is on a separate branch so i'll add that in there.

what specifically is it that doesn't work with other fetch clients?

File: examples/backend-adapters/server/src/effect-platform.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+import { createRouteHandler } from "uploadthing/effect-platform";
+
+import { uploadRouter } from "./router";
+
+const uploadthingRouter = createRouteHandler({
+  router: uploadRouter,
+});
+
+/**
+ * Simple CORS middleware that allows everything...
```
Comment (markflorkowski) [inline examples/backend-adapters/server/src/effect-platform.ts:17]:
```suggestion
 * Simple CORS middleware that allows everything
```

**Review by markflorkowski — APPROVED** [review]
It's very cool how small these adapters can be

**Pull request merged**